packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+
+This interface is what a column web component (i.e. a slotted column) must implement.
+
+```TS
+interface ITableColumn {
+    header?: string; // the string to use in the header for the column
+    headerTemplateFn?: () => ViewTemplate; // the template to use for the header (string if none provided)
+    cellTemplateFn: () => ViewTemplate<ITableCellData>; // the template to use to render the cell content for the column (required)
+    getDataKeys: () => string[]; // the set of keys the table needs to provide the cellTemplateFn with the proper subset of data
+}
+```
+
+The `ITableColumn` interface would be updated to support other features not covered in this HLD such as sorting and grouping.
+
+### `ITableCellData`
+
+In order for the `cellTemplateFn` in the above interface to be able to act on the necessary table data, we should have a type that the template function can work with directly through.
+
+```TS
+type tableData = string | number | boolean | Date | null | undefined;
+
+interface ITableCellData {
+    data: { [key: string]: tableData };
+}
+
+/**
+ * @internal
+ */
+class TableCellData<TData extends { [key: string]: tableData }> implements ITableCellData {
+    public constructor(public data: TData) { }
+}
+```
+
+`TData` here represents a subset of the overall data provided to the table. The `getDataKeys()` method on the `ITableColumn` interface will be used to retrieve the subset of data to provide to a particular `TableCellData` instance.
+
+With these mechanisms in place we should be able to define a series of column providers to handle all of the basic use cases. For example, the `ITableColumn` implementation we could create for rendering data as a read-only `NimbleTextField` could look like this:
+
+```TS
+public TableColumnText extends FoundationElement implements ITableColumn {
+    ...
+
+    @attr
+    public valueKey: string;
+
+    public getDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public cellTemplateFn(): ViewTemplate<ITableCellData> {
+        return html<ITableCellData>`
+            <nimble-text-field readonly="true" value=${x => x.data[this.valueKey]}>
+            </nimble-text-field>
+        `;
+    }
+}
+```
+
+Enabling a particular column to access multiple pieces of data to use in its rendering is equally trivial:
+
+```TS
+public TableColumnNumberWithUnit extends FoundationElement implements ITableColumn {
+    ...
+
+    @attr
+    public valueKey: string;
+
+    @attr
+    public unitKey: string;
+
+    public getDataKeys(): string[] {
+        return [valueKey, unitKey];
+    }
+
+    public cellTemplateFn(): ViewTemplate<ITableCellData> {
+        return html<ITableCellData>`
+            <nimble-text-field readonly="true" value=${x => this.formatData(x.data[this.valueKey], x.data[this.unitKey])}>
+            </nimble-text-field>
+        `;
+    }
+
+    public formatData(value: unknown, unit: unknown): string {
+        return `${(value as string)}, ${(unit as string)}`;
```

Comment:
The `as` casts are interesting to think about. In a dream world we could find a way to restrict you to providing data keys that point to data of a specific type (in this case `value` would be a `number` column and `unit` would be `string` column.

I don't have a complete enough picture to know a) whether it would be possible to verify that at compile time and b) whether that would be unnecessarily restrictive for other use cases. Just want to raise it to see if you've thought about it.

Reply:
This has been addressed in the updated version that uses stricter typing.